lib/ActivityStreams.js

```diff
@@ -132,17 +142,19 @@ ActivityStreams.prototype = {
   },
 
   /**
-   * Broadcast places changes to pages
+   * Handles changes to places
    */
   _handlePlacesChanges(eventName, data) {
-    this.broadcast(am.actions.Response("RECEIVE_PLACES_CHANGES", data));
-  },
-
-  /**
-   * Broadcast bookmark changes to pages
-   */
-  _handleBookmarkChanges(eventName, data) {
-    this.broadcast(am.actions.Response("RECEIVE_BOOKMARKS_CHANGES", data));
+    /* note: this will execute for each of the 3 notifications that occur
+     * when adding a visit: frecency:-1, frecency: real frecency, title */
+    this._invalidatePlacesCache();
+    this._asyncBuildPlacesCache();
+
```

Comment:
I was going to handle this in a later issue. In your outlined scenario, this will actually cause up to 30 history events.
3 for each new link:

1. link creation (with frecency -1)
2. frecency adjustment
3. title change

Also, the solution you propose might not work as a user expects: it will rebuild cache the first time a user creates a new history entry, but potentially not the other times, meaning the results might be stale.

The concurrency scenario you outline will not happen: Javascript is single-threaded. Every synchronous operation will occur one after the other. A sequence of invalidate/build will occur one after the other.